### PR TITLE
[development] refactored the StatisticalProfiler(Proxy) tests and benchmarks

### PR DESCRIPTION
**Pull request by AMZN-ScottR**
- Refactored the unit tests to be more flexible in testing various template types
- Converted and expanded the "performance" tests to use the benchmark framework
- Removed the non-standard `AZStd::shared_spin_mutex` type alias along with the few usages

[Unit Tests Results](https://github.com/o3de/o3de/files/8166059/tests_results.txt)
[Benchmarks Results](https://github.com/o3de/o3de/files/8166056/benchmarks_results.txt)

Signed-off-by: AMZN-ScottR <24445312+AMZN-ScottR@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzCore/Tests/StatisticalProfiler.cpp [hunk 1]
```diff
@@ -5,847 +5,651 @@
...
+
+        ASSERT_TRUE(profiler.GetStatistic(statIdThread4) != nullptr);
+        EXPECT_EQ(profiler.GetStatistic(statIdThread4)->GetNumSamples(), 1);
+        ASSERT_TRUE(profiler.GetStatistic(statIdThread4Loop) != nullptr);
+        EXPECT_EQ(profiler.GetStatistic(statIdThread4Loop)->GetNumSamples(), MediumIterationCount);
 
-        //Clean Up
         proxy->ActivateProfiler(ProfilerProxyGroup, false);
     }
+}//namespace UnitTest
```
Comment (jeremyong-az) [inline Code/Framework/AzCore/Tests/StatisticalProfiler.cpp:413]: Would it be a lot of work to separate the testing code to a different file/lib?

Comment (jeremyong-az) [inline Code/Framework/AzCore/Tests/StatisticalProfiler.cpp:413]: Would it be a lot of work to separate the testing code to a different file and/or lib from the benchmark?

**Review by jeremyong-az — APPROVED** [review]
Thanks for the refactor. Looking at the benchmark code, I'm not completely convinced it is meaningful enough to benchmark the entire matrix of types/counters. Would one or two representative ones be enough? It would cut testing time and still provide pretty much all of the benefit.

**Comment by AMZN-ScottR** [comment]
> Thanks for the refactor. Looking at the benchmark code, I'm not completely convinced it is meaningful enough to benchmark the entire matrix of types/counters. Would one or two representative ones be enough? It would cut testing time and still provide pretty much all of the benefit.

Completely fair benchmarking the entire matrix of tested types is overkill.  I'll go ahead and trim those down.

File: Code/Framework/AzCore/Tests/StatisticalProfiler.cpp (hunk 1 above)
Comment (AMZN-ScottR) [reply Code/Framework/AzCore/Tests/StatisticalProfiler.cpp:413]: Splitting them into different files wouldn't be a problem, different libs would be a completely separate story though.

File: Code/Framework/AzCore/Tests/StatisticalProfiler.cpp [hunk 2]
```diff
@@ -5,847 +5,651 @@
...
+        for (int i = 0; i < loopCount; ++i)
         {
-            CODE_PROFILER_PUSH_TIME(profiler, statNamePerformance)
-                int counter = 0;
-            for (int i = 0; i < iter_count; i++)
-            {
-                CODE_PROFILER_PUSH_TIME(profiler, statNameBlock)
-                    counter++;
-            }
+            typename ProfilerType::TimedScope loopScope(profiler, loopId);
```
[the remark this answers is not in the document]
Comment (AMZN-ScottR) [reply Code/Framework/AzCore/Tests/StatisticalProfiler.cpp:75]: Brace-initialization sadly won't make a difference for using nested types from template inputs.  These instances would need to be replaced with another type alias to remove.

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**